File: Scripts/bigint_generate_minRequiredWidth_tests/generate_numbers.py
import random

# ===========
# === Smi ===
# ===========

smi_max = 2147483647
smi_min = -2147483648


def generate_smi_numbers():
    result = []
    result.append(0)
    result.append(1)
    result.append(-1)

    value = smi_max

    while value != 0:
        result.append(value)
        result.append(-value)
        value = int(value / 2)

    return result

# ============
# === Heap ===
# ============


word_max = 18446744073709551615
word_min = 0


def generate_heap_numbers(count_but_not_really):
    result = []

    result.append(word_max)
    result.append(-word_max)

    maxWordCount = 3

    for i in range(0, count_but_not_really):
        min_1_word_because_we_already_added_0 = 1
        word_count = (i % maxWordCount) + min_1_word_because_we_already_added_0

        value = 1
        for j in range(0, word_count):
            word = random.randint(0, word_max)
            value = value * word_max + word
            word += 1

        # Hmm... this will produce very boring numbers
        # Let's make it a little bit more random:
        removeCount = random.randint(0, 10)
        s = str(value)[removeCount:]
        value = int(s)

        assert value > smi_max

        result.append(value)
        result.append(-value)

    return result

File: Tests/BigIntTests/Helpers/GenerateNumbers.swift
@testable import BigInt

private typealias Word = BigIntStorage.Word

private let smiZero = Smi.Storage.zero
private let smiMax = Smi.Storage.max
private let smiMin = Smi.Storage.min

// MARK: - Fixed width integers

/// Will return `2 * countButNotReally + 3` values (don't ask).
internal func generateSmiValues(countButNotReally: Int) -> [Smi.Storage] {
  return generateValues(
    countButNotReally: countButNotReally,
    type: Smi.Storage.self
  )
}

/// Will return `2 * countButNotReally + 3` values (don't ask).
internal func generateIntValues(countButNotReally: Int) -> [Int] {
  return generateValues(
    countButNotReally: countButNotReally,
    type: Int.self
  )
}

private func generateValues<T: FixedWidthInteger>(
  countButNotReally: Int,
  type: T.Type
) -> [T] {
  assert(countButNotReally > 0)

  var result = [T]()
  result.append(0)
  result.append(1)
  result.append(-1)

  let step = Int(T.max) / countButNotReally

  // 1st iteration will append 'T.min' and 'T.max'
  for i in 0..<countButNotReally {
    let s = i * step

    let fromMax = T(Int(T.max) - s)
    result.append(fromMax)

    let fromMin = T(Int(T.min) + s)
    result.append(fromMin)
  }

  return result
}

// MARK: - Heap

internal struct HeapPrototype {

  internal let isNegative: Bool
  internal let words: [BigIntStorage.Word]

  internal var isPositive: Bool {
    return !self.isNegative
  }

  internal var isZero: Bool {
    return self.words.isEmpty
  }

  internal var hasMagnitudeOfOne: Bool {
    return self.words.count == 1 && self.words[0] == 1
  }

  internal func create() -> BigIntHeap {
    return BigIntHeap(isNegative: self.isNegative, words: self.words)
  }
}

/// Will return `2 * countButNotReally + 5` values (don't ask).
///
/// We do not return `BigIntHeap` directly because in some cases
/// (for example equality tests) you may want to create more than 1 value
/// INDEPENDENTLY.
internal func generateHeapValues(countButNotReally: Int,
                                 maxWordCount: Int = 3) -> [HeapPrototype] {
  var result = [HeapPrototype]()
  result.append(HeapPrototype(isNegative: false, words: [])) //  0
  result.append(HeapPrototype(isNegative: false, words: [1])) //  1
  result.append(HeapPrototype(isNegative: true, words: [1])) // -1
  result.append(HeapPrototype(isNegative: false, words: [.max])) //  Word.max
  result.append(HeapPrototype(isNegative: true, words: [.max])) // -Word.max

  var word = Word(2) // Start from '2' and go up
  for i in 0..<countButNotReally {
    let min1WordBecauseWeAlreadyAddedZero = 1
    let wordCount = (i % maxWordCount) + min1WordBecauseWeAlreadyAddedZero

    var words = [Word]()
    for _ in 0..<wordCount {
      words.append(word)
      word += 1
    }

    result.append(HeapPrototype(isNegative: false, words: words))
    result.append(HeapPrototype(isNegative: true, words: words))
  }

  return result
}
